### PR TITLE
[ci skip] Improve accessibility

### PR DESCRIPTION
**Pull request by SleeplessByte**
### Motivation / Background

As requested by @AmandaPerino, this PR aims to improve the accessibility of the guides. It continues on the work in #51341 #51372 #51385 and #51444. It also addresses whatever wasn't addressed from https://github.com/rails/rails/pull/51341#issuecomment-2007717282.

### Detail

Each commit in the PR has a descriptive title and often a message why the change was made, which I recommend we keep when merging instead of squashing it. Screenshots are listed below. A few of the commits that introduce fixes or features are listed here:

* [Add label and accessible content](https://github.com/rails/rails/commit/4f2b25ba54d82bc897fb33f681214f5bf05cb0bd) 
  
  Having a select element without a label outside a form isn't directly problematic, but the text Version is the label of the select and should act as such.

  Because the select element is set up to browse to a different page which loses the current focus position, users must be instructed about this context change. It would be much better to display this content to everyone, but having this for screenreaders only is better than not having it at all.

* [Fix accessible text for voice commands](https://github.com/rails/rails/commit/1863e5590cecac6ecfd0c4c031f1c4c967d0d768) 
  
  Using the title attribute means some screenreaders will not be able to show the expanded accessible text at all, but using aria-label loses the hover text currently available.

  If the title attribute is still used, the accessible label should start with the same content as the visible label. This is necessary for most voice operated tooling which only match starting text.

  Having "Return" as the start of the accessible label and "Guides" as the start of the visible label means in most cases users that rely on voice commands will not be able to browse here.

* [Add landmarks to all guide pages](https://github.com/rails/rails/commit/9669f4be012f34488d52e16d0b6e40a11399e414) 
  
  Landmarks such as header, section, footer, article, and the main content landmark help navigating when using assistive tooling.

  Since each guide is a self-contained article, a header section with the guide introduction and navigation, a footer section which is not part of the content but a call to action, and the rest can be clearly separated, landmarks here are appropriate.

* [Make back-to-top work without JavaScript](https://github.com/rails/rails/commit/b33278c01384a7612d0adf2270456718181b5acb) 

  By doing feature detection on animation timeline CSS, the graceful degradation can take care of the visibility of the back-to-top button. In case no JavaScript is available, it should always be shown.

* [Enable back-to-top without JavaScript](https://github.com/rails/rails/commit/430f27c959f0df30b6baeea3d688d8fa5448f842) 

  And add an accessible label to the button. It would be better to just show the text back to top underneath the arrow, but this change is again better than not having it at all.

* [Consistent scrolling behaviour respecting pref](https://github.com/rails/rails/commit/f2ae0318736a286a13d896cad8527bd2d1443c50) 

  Use the preference to detect if a user wants reduced motion and if not enable smooth scrolling. This takes care of the back-to-top smooth scrolling, and also adds it to anchor links. If this is unwanted, the change can be reverted and a matchmedia check should be done in JavaScript instead before smooth scrolling

* [Add skip links to quickly navigate past navs](https://github.com/rails/rails/commit/653b292ac26b2df59ebeb44fc9415ebc8130243b) 
  
  Skip links are a common practice to allow keyboard and other users to skip past large blocks of content that are present on every page, which in this case are the secondary navigation at the top,
the version selector, and the chapter navigation.

  The skip links only show when focused.

* [Wrap tables to make them responsive](https://github.com/rails/rails/commit/e2cfe8fdfccfcb47aec2f358b0badab79c817981)

* [Make back-to-top link to skip link](https://github.com/rails/rails/commit/ad966a42de9e166ab33026b28fc8dda64f5fbdd1) 

  This allow screen readers to announce something, because linking to an empty fragment or the body element doesn't announce anything useful.

* [Add branded selectable color](https://github.com/rails/rails/commit/2557614d872629462e354b2bd1536736aa2077be) 

  Fixes selecting text in dark mode and brands it for both modes.

* [Make more info and guides index accessible](https://github.com/rails/rails/commit/f9eaeb18d8c5712ae0f2edf351ff9829c996bf78) 

  * Adds aria-controls, which binds the control to the content.
  * Adds aria-expanded, which signals the current expansion state.
  * Adds focus handling, which focuses the first link or the button
    depending on the expansion state.
  * Adds keyboard closing for the more info button.
  * Adds outside click tracking to close the expanded content.
  * Fixes the more info button to not be focusable.

* [Fix dark mode for guide selector](https://github.com/rails/rails/commit/9c4cfd00492a2bd55db9ed5b5a9a616a962c153a)

* [Add accessible label for guide navigator](https://github.com/rails/rails/commit/93f9819c62890301f398e752b6083c9d2c9f6c40)

### Additional information

Please note that this PR doesn't necessarily make the guides completely accessible and we should continue to improve on this, but this gets us (a big) part of the way there, together with the changes that were already made prior.

| Before | After |
| ------ | ----- |
| You could not focus the selector pressing "Version" | a11y-focusable-label-version |
| The element (version selector) was broken if JavaScript was disabled | a11y-no-js-desktop |
| The element (version selector) was broken if JavaScript was disabled | a11y-no-js-mobile |
| a11y-before-table-wrap | a11y-after-table-wrap |
| You could not skip past the secondary top navigation | a11y-main-skip-link |
| You could not skip past the chapter navigation | a11y-chapter-skip-link |
| a11y-selectable-text-before | a11y-selectable-text |
| There was no label indicating what the select does | a11y-guide-selector |

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: guides/assets/javascripts/guides.js
Comment (p8) [inline guides/assets/javascripts/guides.js:34]: While we are renaming things here, I think we can probably also rename the id of the button and rename the `guides` to `guidesMenu`?
```suggestion
    var guidesMenuButton = document.getElementById("guidesMenuButton");
    var guidesMenu = document.getElementById(guidesMenuButton.getAttribute('aria-controls'));
```

Comment (SleeplessByte) [reply guides/assets/javascripts/guides.js:34]: Sure!

I tried not to touch unrelated things except for those that really didn't make sense in context.

Any other renames we should apply right away?

Comment (p8) [reply guides/assets/javascripts/guides.js:34]: I think these are the most obvious ones that are touched by this PR.
Any others I'm missing?

Comment (SleeplessByte) [reply guides/assets/javascripts/guides.js:34]: I'd opt to rename container to main as it's... not a container. s-hidden also doesn't really mean anything and there is already a different class used to hide things conditionally.

Should I just make changes to improve the maintainability now that we touch these things?

**Comment by jathayde** [comment]
This looks great @SleeplessByte. I agree on the naming `container` to something else. Open to any improvements there. I was probably overly sensitive in not rewriting it wholesale, so a lot of these things persisted when I should have pulled them.

**Comment by SleeplessByte** [comment]
On the other hand it does add a bit of noise to the PR. I understand the rethoric in the contributing guide hehe.

I'll wait for @carlosantoniodasilva review before applying to keep it as noise free.

File: guides/source/layout.html.erb
Comment (brunoprietog) [inline guides/source/layout.html.erb:74]: What do you think about giving a button role to this link? It's not really navigating to any page, and that's the only purpose of a link actually. (Unless JavaScript is disabled, in which case it shouldn't have those other ARIA attributes either if that's the case).

```suggestion
            <a href="index.html" id="guidesMenu" role="button" aria-controls="guides" aria-expanded="false" class="guides-index-item nav-item">Guides Index</a>
```

File: guides/source/index.html.erb
Comment (brunoprietog) [inline guides/source/index.html.erb:14]: This section is being read as if it were a list of items, and it's not. Let's ignore it.

```suggestion
  <dl role="presentation">
```

**Review by brunoprietog — COMMENTED** [review]
Great work! Thank you so much for working on this. I left a few comments, several have a bias from me as a screen reader user.

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:130]: In html 5, footer may also be an element inside an article or section. I experience the same behaviour as you (because I also use NVDA when I use a screen reader) but I believe jaws does denote it differently.

File: guides/rails_guides/markdown.rb
Comment (SleeplessByte) [reply guides/rails_guides/markdown.rb:155]: What does this accomplish?

I thought same-page anchors were fixed in turbo 😁

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:131]: I think it should!

File: guides/source/index.html.erb
Comment (SleeplessByte) [reply guides/source/index.html.erb:9]: Yeah I agree. It believe the choice was made to have it as nav because on articles it does contain links but on the index it doesn't.

Let's change it here.

File: guides/source/index.html.erb
Comment (SleeplessByte) [reply guides/source/index.html.erb:11]: Same as above

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:40]: It must have a label because it's not the primary navigation (otherwise it fails a wcag criterium), as there is already a primary navigation element.

I hate the label "secondary" personally, do you have a different suggestion?

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:74]: I like it. 

Let's apply it in JavaScript so it's only a button if... It's a button 😁

**Comment by SleeplessByte** [comment]
Thank you @brunoprietog! Very very helpful.

I left a few questions on the review and will apply all other changes.

File: guides/rails_guides/markdown.rb
Comment (brunoprietog) [reply guides/rails_guides/markdown.rb:155]: At least with NVDA it doesn't focus the region correctly 😢

File: guides/source/layout.html.erb
Comment (brunoprietog) [reply guides/source/layout.html.erb:40]: I understand that it's not the primary navigation, but when this element is shown, the element containing the primary navigation is hidden, so in theory this is the one that now becomes the primary navigation element. What do you think?

File: guides/source/layout.html.erb
Comment (brunoprietog) [reply guides/source/layout.html.erb:62]: Well, so I think it's fine to leave it as it's. I didn't really have a very strong opinion on that either.

File: guides/rails_guides/markdown.rb
Comment (SleeplessByte) [reply guides/rails_guides/markdown.rb:155]: Oh weird. It does on mine, but that's good enough reason. I'll make the change!

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:40]: I think you are right. Okay my proposal will be to drop the label and instead add an ERB comment for the next developer that the navs are mutually exclusive and thus not actually secondary.

Comment (jathayde) [reply guides/source/layout.html.erb:40]: Is the name something used for voice navigation or is it more to provide context? Perhaps "global" and "chapter contents" or some other delineation that alludes to the content?

File: guides/source/layout.html.erb
Comment (brunoprietog) [reply guides/source/layout.html.erb:130]: Oh, I don't know. I'm not too convinced about adding so many additional regions beyond the navigation, main region and footer. It adds a lot of noise that is already covered by the heading levels and becomes a bit repetitive information.

By the way, I personally don't tend to use many links to skip to specific sections and hope this can be done natively. And really, there's no way to skip to the body of the article with a command to go to a region. You can easily use 2, but that makes me think we're not being very consistent with using regions, or if we're even using them properly.

What do you think?

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:40]: In our screen readers it will do something along the lines of: "Secondary navigation". It must provide context. However, the "primary" (top) navigation is allowed to not have a contextual label. That's basically what we were discussing here.

File: guides/assets/javascripts/guides.js
Comment (SleeplessByte) [reply guides/assets/javascripts/guides.js:71]: In dabdb0aeef

File: guides/rails_guides/markdown.rb
Comment (SleeplessByte) [reply guides/rails_guides/markdown.rb:159]: In 7437e861a1

File: guides/rails_guides/markdown.rb
Comment (SleeplessByte) [reply guides/rails_guides/markdown.rb:160]: In 25057200c4

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:130]: My personal thoughts here were:

- I am browsing the guides extensively
- I am in the middle of the article and find something I want to give feedback on
- I know there is a _region_ (labelled section) that explains how to give feedback
- I want to go there straight away

With NVDA I can get there right away when browsing the article pressing <kbd>D</kbd>, which should navigate to the next landmark or region. In this case the footer is nested inside article, and thus not _complementary_, but it is labelled by the heading, and thus it becomes a section with role="region", iirc. Using the TOC navigation here would be sub-optimal as it doesn't allow be to go to "last heading of the current level", by pressing <kbd>2</kbd>. I would need to press it repeatedly, right? Since there are so little regions, that's not the case.

Does this make sense?

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:40]: @brunoprietog I actually think this can be made just a `div`. It's not a `nav` in the sense that it only holds a single button that expands a set of links. Do you agree?

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:130]: We have a region for the article: https://github.com/rails/rails/pull/51499/files/93f9819c62890301f398e752b6083c9d2c9f6c40#diff-dc279d925a05373ac5f34307441799f98ac4accbe0b0a8f12f99f703585afa7dR113 Is that not working for you? I haven't tested mine but I can.

It _should_ be:

```html
<main>
  <article>

    <header id="feature">
      article introduction

      chapter index / toc
    </header>   

    article contents with lots of headings    

    <footer>
     <h2>Feedback</h2>

     feedback explanation
    </footer>
  </article>
</main>
```

Comment (SleeplessByte) [reply guides/source/layout.html.erb:130]: (sidenote: I am fine with explicit role-ing things if SRs don't recognise it, so let's do that if we must)

Comment (brunoprietog) [reply guides/source/layout.html.erb:130]: Right! So I think assigning the explicit roles would be enough to solve this.

In any case, an article is not exposed as a region, it's exposed as another type of element. In fact, at least in NVDA, there's no default shortcut to jump to it, I configured the A to do so but it's not the common case.

Comment (SleeplessByte) [reply guides/source/layout.html.erb:130]: Cool! Yeah, I _love_ different support for different SRs and even different versions of the same SR 🗡️ 😁. 

I'll explicitly role both the article header and article footer (so your NVDA can skip back/forth to them)

Comment (brunoprietog) [reply guides/source/layout.html.erb:130]: Yes, sadly. Just like the history of browser feature compatibility 😰

File: guides/assets/stylesrc/_dark.scss
Comment (p8) [inline guides/assets/stylesrc/_dark.scss:108]:
```suggestion
    #mobile-navigation-bar {
      background-color: $gray-900;
```

Comment (SleeplessByte) [reply guides/assets/stylesrc/_dark.scss:108]: (something something we should use an auto-formatter imho hehe)

Comment (p8) [reply guides/assets/stylesrc/_dark.scss:108]: Yes :)

File: guides/source/layout.html.erb
Comment (SleeplessByte) [reply guides/source/layout.html.erb:124]: For the guides index it would have the entire index in there, which kinda makes sense. Would we label the region? Or just as a way to jump there? It is harder if we need to label it, but still possible.

Comment (brunoprietog) [reply guides/source/layout.html.erb:124]: I mean, add a region to the article body wrapper so that you can easily jump to it using quick navigation commands between regions. But only in the article body, not at the general level, because in the guide index there should be no such region.

We should probably create an extra wrapper in `markdown.rb`.

Comment (SleeplessByte) [reply guides/source/layout.html.erb:124]: The change is

```ruby
doc.wrap("<section role='region'>")
```

on line 134, and we need some changes to the anchors checking then. However, I can not make this change because after rebasing I can no longer run `rake guides:generate` probably because it expects the latest `rails` version which I cannot install on the box I am running this PR on. I therefore would like to defer this change for now.

**Comment by SleeplessByte** [comment]
@AmandaPerino ready for final review

**Comment by carlosantoniodasilva** [comment]
@SleeplessByte what have you used to auto-format the JS/SCSS files? The extra formatting is making it a bit more difficult to separate what's actual change vs not, so I'm wondering if we should push a formatting change first, and merging onto yours to make the diff less noisy.

**Comment by SleeplessByte** [comment]
It was me accidentally not turning off format on save.

It's using the latest prettier. You can run once using `npx prettier --write .` inside the guides directory with default settings.

Apologies for that!

**Comment by jathayde** [comment]
Is that a code standard we ascribe to across the Rails repo? I'm fine if it is, but we need to run that on everything, and then make it part of a check in process. The underscores for IDs, hyphens for classes is something I've done for 20 years. Not saying it's right, but if we are going to implement a standard, let's figure out what that is and make part of the checkboxes on PRs, etc.

**Comment by carlosantoniodasilva** [comment]
@jathayde Rails does not have any sort of standard for those, as it's just JS/CSS used for the guides themselves. (it does have standards for the actual framework Ruby code, however.)

I assumed the hyphen changes were related to the accessibility, but I haven't gotten that far as looking into the individual commits to verify my assumption. Either way, I'd rather not make a commitment of running any automated formatting tools and come up with too many standards for the assets part of the guides, it's supposed to be a lightweight process, focused on the content. (that said, I don't mind running a formatting tool like prettier every once in a while, as it's what most people use as default for these anyway these days.)

**Comment by SleeplessByte** [comment]
I'll pull in latest changes from #51504

**Comment by zzak** [comment]
Thanks for your work on this!

Just putting it out there but there is already a linter and w3c validator (not currently run in CI, I believe) on the guides:
https://github.com/rails/rails/blob/965b8c372ac0e19021f92cf798095af6046ab140/guides/Rakefile#L26-L49

[[example](https://buildkite.com/rails/rails/builds/107333#018f88f6-630f-4f87-b183-14dfdb5d06e2)]

So there if there was a lightweight way to validate the a11y too, I don't see why we wouldn't take advantage of that and prevent regressions from the work done here.

**Comment by zzak** [comment]
I don't want to make more work for you or reviewers here, apologies for the digression.
Let's chat about it after this gets merged.

**Comment by brunoprietog** [comment]
I still want to second @SleeplessByte, automated tests might help with some minor use cases but they rarely detect serious accessibility issues.

**Comment by SleeplessByte** [comment]
I have not forgotten about this. Will attempt to finish this weekend now that the other PRs are merged.

**Comment by SleeplessByte** [comment]
I will do a final test with my screen reader tomorrow (Monday) and then it should be good for final review.

**Comment by SleeplessByte** [comment]
RTL implementation of the highlight feature

The nav highlight is now also supported in RTL mode, shouldn't be flaky with flick/fast scrolling, and should now work both going up and down. 

@AmandaPerino ready for final review.

**Comment by p8** [comment]
Thanks for sticking with this @SleeplessByte !
It seems the side bar no longer scrolls when the active element is outside the visible region:
https://github.com/rails/rails/assets/28561/ca82b356-e8b7-4223-93a8-3c42b4b1ce40

**Comment by p8** [comment]
Is the red highlight maybe a bit too much in light mode?

Current version:
image

This PR:
image

**Comment by SleeplessByte** [comment]
> Thanks for sticking with this @SleeplessByte !
> It seems the side bar no longer scrolls when the nav highlight is outside the visible region:
> https://github.com/rails/rails/assets/28561/ca82b356-e8b7-4223-93a8-3c42b4b1ce40
> 
> I'm also getting the following error (maybe related):
> > Uncaught ReferenceError: backToTop is not defined
>     toggleBackToTop .../rails/guides/output/javascripts/guides.js:113
> 
> 

Yes, the scrolling of the sidebar stops scrolling on windows, so we need to implement that differently but i want to get this over the line before we do. See the comment in the code.

The second thing is a merge artifact. Will fix!

**Comment by SleeplessByte** [comment]
> Is the white text with red highlight maybe a bit too much in light mode?
> 
> Current version:
> image
> 
> This PR:
> image
> 
> 
> 
> 

Because the links are red, I dont think it's smart to invert it, but an off-white may help to reduce the contrast a bit.

**Comment by p8** [comment]
> Yes, the scrolling of the sidebar (`scrollTo`, `scrollTop` etc), stops scrolling on Windows, which means that as you are scrolling the main element, and the side element is scrolled, it completely stops scrolling the main element. So we need to implement that differently but I want to get this PR over the line before we do. See the comment in the code to understand more.

This happens to me on OSX, both in FireFox and Chrome. I'm not allowed to merge code (only documentation) but I'd be a bit hesitant to merge this while the scrolling of the sidebar is a bit broken.

I know you put a lot of time and energy into this and this is all volunteer work.
Not sure how comfortable you are with git and I don't want put more work on your plate.
Would it be possible to split the sidebar changes from the other changes?
Or could we easily cherry-pick some of the earlier commits?

> The second thing is a merge artifact. Will fix!

Thanks. This has been fixed.

**Comment by SleeplessByte** [comment]
This is broken on the live site / main if you have smooth scrolling on, so this PR does not break it.

I'm okay with trying to find a solution to the issue instead if that is what's keeping this from being merged. Why don't I do that before cherry picking the commits (which I also know how to do).

**Comment by SleeplessByte** [comment]
(to respond to my comment about windows: I meant the implementation before this PR.)

**Comment by p8** [comment]
> (to respond to my comment about windows: I meant the implementation before this PR.)

Ah, that makes sense. Thanks!

**Comment by SleeplessByte** [comment]
Still on the list, I think I have a solution. Will try to test it some more this week.

**Comment by p8** [comment]
@SleeplessByte It would be nice if we could merge some of these changes.
We are getting [PR's ](https://github.com/rails/rails/pull/52838)to fix issues addressed by this PR.

**Comment by SleeplessByte** [comment]
Yap.

Ping me again on Friday if I haven't pushed yet.

**Comment by p8** [comment]
Thanks @SleeplessByte !

**Comment by SleeplessByte** [comment]
Was able to empty my friday out so I'll work on this tomorrow!

**Comment by SleeplessByte** [comment]
@p8 your concern has been addressed.

**Comment by p8** [comment]
Thanks @SleeplessByte, this is looking great!
@carlosantoniodasilva do you think this could be merged?

**Review by p8 — APPROVED** [review]

**Comment by p8** [comment]
@zzak Do you think this is good to go?

**Comment by zzak** [comment]
@SleeplessByte Could you squash your commits please? I will try to make time to review this soon. :bow:

**Comment by SleeplessByte** [comment]
@zzak squashing will lose a lot of valuable information for anyone later looking at the individual lines of code plus git blame. Almost each commit has a detailed message of the whys.

If possible I'd rather not, so to make future maintenance easier.

**Comment by zzak** [comment]
Ok, I can see you went to great lengths to provide additional context in each commit, and since this is not touching Rails framework code it maybe acceptable as-is but our policy is to always squash commits before merging.

I'm still catching up on stuff, have been sick so apologies for the delay. Another good person to ask might be @seanpdoyle if they can give this a look :bow:

**Comment by SleeplessByte** [comment]
> our policy is to always squash commits before merging.

Yep, hence me asking for the exception in the original PR message 🗡️ 
Hoping this can be looked at soon though, I think these changes are impactful for many users.

**Comment by zzak** [comment]
Ok I had some time to at least check the changes listed above, especially the darkmode and screen reader stuff, definitely an improvement on the UX.

Two things stood out to me.

Screenshot from 2024-10-18 20-13-41

The "skip to article body" link covers the nav and it feels like a bug.

Screenshot from 2024-10-18 20-21-35

The guides selector is blank when viewing on a page in a small browser window.

I didn't take a hard look at the code, just checking if there were any obvious traps. I'm not a big fan of the cosmetic changes, like whitespace, semi-colons, etc -- just because it makes the diff much larger than it needs to be. These types of changes should be delegated to a linter which we can easily auto-format for us and get everyone on the same page -- but we haven't needed one for the guides until now so I'm not sure. There are also multiple "merge commits" on the branch that I'd like to see rebased out, but these are just some very minor knits.

**Review by seanpdoyle — APPROVED** [review]
I agree with @zzak that the volume of changes related to indentation, the addition of `;` characters, and changes from `_` characters to `-` characters made reviewing this diff more challenging than it would be otherwise. Having said that, viewing the diff [ignoring spacing changes](https://github.com/rails/rails/pull/51499/files?w=1) improved that experience.

Changes to the document's structure, elements' semantics, and form controls' accessible names are all improvements. I'm less sure about the particulars of the JavaScript implementation, but that feedback might be better reserved for a separate pull request discussion.

**Comment by SleeplessByte** [comment]
> The "skip to article body" link covers the nav and it feels like a bug.

It's common practice. I am personally very used to skiplinks. However, given that this can be entirely controlled by CSS, perhaps an improvement could be addressed in a future PR?

> The guides selector is blank when viewing on a page in a small browser window.

I saw that too. If okay, I'd like to address that in a follow-up PR myself.

> I'm not a big fan of the cosmetic changes, [...]

Wholeheartedly agree with everything you said here @zzak. In fact, in my response to an earlier review on this PR I said the same. This could all be prevented if there was an auto-formatter (not linter). The wide variation of styles in the same JavaScript file threw me off a lot when trying to dissect how to change it. Personally I do think you needed it, given the disjoint style it is in on main. My two cents? It should be added! And then we can all agree that we don't agree, but the system decides, and auto-formatting wins all!

> There are also multiple "merge commits" on the branch that I'd like to see rebased out

I've rebased this PR 6 times now, and they are not easy because of a lot of conflicts with `main`. 
The "merge" commits you see (https://github.com/rails/rails/pull/51499/commits/43b8eb88df42309647c28b8ca59570163aa7a410 and https://github.com/rails/rails/pull/51499/commits/3d389b01c0ec534b465de836ab723fa16b9b4324) are likely conflict resolution commits. TBH they're only here because this has been kept on the shelve since I first did the work for so long. If this is an absolute must, I can do it, but if the PR then sits for another few months accumulating conflicts, I'll likely no longer be interested, as I don't know if I can find the will to go through the interactive process again. If one of the maintainers happens to be a star with pulling commits out (without the full rebase), this PR *is* open for edits! 

> Having said that, viewing the diff [ignoring spacing changes](https://github.com/rails/rails/pull/51499/files?w=1) improved that experience.

Great! I tend to use the ignore whitespace setting myself as well. I don't understand the `;` comment though. I've followed the style of the file which was semis yes, so the only places where it now has an `;` it was missing it before. As for `-` vs `_` in the names/ids, I tried to be consistent. It's discussed earlier in the PR. My screenreader separates words better when hyphenated than when underscored, but I do realise this feels like noise to the PR.

----------------------------------------------------

I am fine with a follow-up issue I'm assigned to to address issues or improvements (in smaller bits, so easier to review perhaps?) if that's something that would help get this merged.

**Comment by Earlopain** [comment]
I believe this PR has caused issues on smaller screens with the chapter select: https://github.com/rails/rails/issues/53500. When the chapter select changes to take up the entire screen width, it is impossible to scroll down.

**Comment by brunoprietog** [comment]
Actually, the reason behind using the JavaScript code there was because we only wanted to set the roles when we have JavaScript enabled. If that's not a concern and we'll assume that will always be the case, it's probably fine.

**Pull request merged**
